### ISSUE TITLE
[RAG-Quality] RAPTOR — hiyerarşik özet chunk ağacı ile çok katmanlı retrieval

### ISSUE THREAD

**Issue by defectGI** [defectGI/tepegoz#181]
## Motivasyon

Tepegoz'un chunking'i paragraf düzeyinde çalışıyor — her chunk bağımsız bir metin parçası. Bu, "dokümanın genel konusu ne?" veya "X ve Y konularını karşılaştır" gibi geniş kapsamlı sorularda başarısız oluyor çünkü cevap tek chunk'a sığmıyor. Outline node'lar bu boşluğu kısmen kapatıyor ama sadece başlık düzeyinde — içerik özeti yok.

RAPTOR (Recursive Abstractive Processing for Tree-Organized Retrieval), chunk'ları benzerlik bazlı gruplar ve her grubu LLM ile özetleyip üst katman chunk'lar oluşturuyor. Bu özetler de embed edilip aranabilir hale geliyor → hem detaylı hem genel sorular aynı pipeline'da cevaplanıyor.

## RAGFlow Referansı

RAGFlow'un `rag/raptor.py` — `RecursiveAbstractiveProcessing4TreeOrganizedRetrieval` sınıfı.

### Temel Algoritma (Classic RAPTOR, satır 642-754)
```
while chunk sayısı > 1:
    1. Chunk embedding'lerini UMAP ile boyut azalt (12 dim, cosine)
    2. GMM veya AHC ile benzer chunk'ları kümele
    3. Her kümeyi LLM ile özetle
    4. Özeti embed et → yeni üst katman chunk
    5. Tekrarla → piramit oluşana kadar
```

### Kümeleme Stratejileri
- **GMM (Gaussian Mixture Model)**: BIC skoruyla optimal küme sayısı, `threshold=0.1` ile soft assignment
- **AHC (Agglomerative Hierarchical Clustering)**: Ward linkage, dendrogram gap heuristic ile küme sayısı, ardından centroid reassignment ile refinement (`_adjust_tree_nodes`)

### Psi Tree Builder (Alternatif, satır 556-573)
- Union-Find tabanlı merge tree — daha dengeli yapı
- Büyük chunk setleri için bucket splitting (cosine-based k-means)
- Leaf pair ranking: tüm çiftleri cosine similarity ile sırala, en benzer çiftleri birleştir
- Rebalance: max_cluster fanout aşılırsa otomatik düzeltme

### Özet Oluşturma (satır 302-343)
```python
async def _summarize_texts(self, texts, callback, task_id):
    cluster_content = "\n".join([truncate(t, len_per_chunk) for t in texts])
    cnt = await self._chat("You're a helpful assistant.",
        [{"role": "user", "content": self._prompt.format(cluster_content=cluster_content)}],
        {"max_tokens": max(self._max_token, 512)})
    embds = await self._embedding_encode(cnt)
    return cnt, embds
```
- Her küme LLM'e gönderilip özetleniyor
- Özet embed ediliyor → üst katman chunk olarak ekleniyor
- Embedding ve LLM sonuçları cache'leniyor
- Hata toleransı: `max_errors=3` — birkaç küme başarısız olsa bile devam ediyor
- Task cancellation desteği — uzun süren işlemler iptal edilebiliyor

### Katman Yönetimi
```python
layers = [(0, len(chunks))]  # [(start_idx, end_idx), ...]
```
- Her katmanın chunk aralığı takip ediliyor
- Retrieval sırasında tüm katmanlar aranıyor — detaylı ve özet chunk'lar birlikte sıralanıyor

## Tepegoz'da Mevcut Durum

- `backend/app/services/rag/chunker.py`: Paragraf bazlı chunking, token budget ile
- `outline_nodes` collection: Başlık embedding'leri (dense-only) — yapısal navigasyon
- `doc_records` collection: Doküman metadata
- LLM erişimi var (`llm_service.chat_completion`) — özetleme için kullanılabilir
- `embedding_service.embed()`: Dense + sparse embedding — özet chunk'lar için kullanılabilir

### Faz 1: RAPTOR Core Modülü
- [ ] `backend/app/services/rag/raptor.py` modülü oluştur
- [ ] `RaptorSummarizer` sınıfı:
  - `__init__(max_cluster, max_token, threshold, clustering_method)`
  - `async def build_summary_layers(chunks: list[tuple[str, list[float]]]) -> list[tuple[str, list[float]]]`
- [ ] Kümeleme: sklearn `AgglomerativeClustering` (Ward linkage) — GMM'den daha stabil, RAGFlow'un AHC path'i referans
  - UMAP boyut azaltma (12 dim, cosine metric)
  - Dendrogram gap heuristic ile küme sayısı otomatiği
- [ ] Özetleme: `llm_service.chat_completion` ile küme içeriğini özetle
  - Prompt: "Aşağıdaki metinlerin kısa ve kapsamlı bir özetini yaz. Önemli teknik detayları koru."
  - `max_tokens=512` — özet chunk'lar orijinallerden kısa olmalı
  - Türkçe/İngilizce — özet kaynak dilin dilinde olmalı
- [ ] Embedding: `embedding_service.embed()` ile özet chunk'ı embed et (dense + sparse)
- [ ] Hata toleransı: Küme özetleme başarısız olursa atla, devam et (RAGFlow'un `max_errors` pattern'i)

### Faz 2: Indexing Entegrasyonu
- [ ] `vector_index.py`'de `index_chunks` sonrası opsiyonel RAPTOR katmanı
- [ ] Özet chunk'ları `company_docs` collection'ına `chunk_type="raptor_summary"` ve `raptor_layer=N` payload'ı ile upsert et
- [ ] Aynı `access_level` + `doc_id` ile — erişim kontrolü korunsun
- [ ] Reindex sırasında eski RAPTOR chunk'ları da silinmeli (`delete_points_by_doc_id` zaten bunu yapıyor)
- [ ] `RAG_RAPTOR_ENABLED=false` env var (default: false — LLM call maliyeti yüksek)

### Faz 3: Retrieval Entegrasyonu
- [ ] RAPTOR chunk'ları normal retrieval'da otomatik aranıyor (aynı collection'da)
- [ ] Opsiyonel: `raptor_layer` filtresi ile sadece özet veya sadece detay chunk'lar aranabilir
- [ ] Query classification (#179) ile entegrasyon:
  - SUMMARY tipi sorgularda `raptor_layer >= 1` filtresi → üst katman özetleri tercih et
  - FACTUAL tipi sorgularda `raptor_layer == 0` filtresi → detay chunk'ları tercih et

### Faz 4: Performans ve Konfigürasyon
- [ ] Bağımlılıklar: `umap-learn`, `scikit-learn` (sklearn zaten var mı kontrol et)
- [ ] `.env.example`'a ekle: `RAG_RAPTOR_ENABLED`, `RAG_RAPTOR_MAX_CLUSTER`, `RAG_RAPTOR_MAX_TOKEN`
- [ ] RAPTOR sadece yeni index veya reindex sırasında çalışmalı — mevcut chunk'lar değişmez
- [ ] Büyük dokümanlar için async/concurrent özetleme (RAGFlow'un `asyncio.gather` pattern'i)

## Bağımlılıklar
- scikit-learn (muhtemelen zaten var — kontrol et)
- umap-learn (yeni bağımlılık)
- numpy (zaten var)

## Risk ve Trade-off
- **LLM maliyeti**: Her doküman indexing'inde N adet LLM call — default off olmalı
- **Index süresi**: Büyük dokümanlar için dakikalar sürebilir — async + progress callback
- **Index boyutu**: Özet chunk'lar collection boyutunu %20-50 artırır
- **CPU torch uyumluluğu**: umap-learn ve sklearn CPU-only — backend'in GPU-free constraint'i ile uyumlu

## Kalite Kontrol
- [ ] RAPTOR modülü birim testleri (mock LLM ile kümeleme + katman oluşturma)
- [ ] Entegrasyon testi: küçük doküman → RAPTOR → retrieval → özet chunk döner
- [ ] `RAG_RAPTOR_ENABLED=false` ile mevcut davranış korunur
- [ ] `make test` geçiyor
- [ ] CPU-only wheel'ler — CUDA bağımlılığı yok

---

## Tasarım Kararları — 2026-05-31 (mimari tartışma sonucu)

Bu bölüm, RAPTOR'un Tepegoz'un mevcut mimarisine — doküman-düzeyi ACL, audit-grade verbatim citation, per-doc content-hash reindex, 7-tool agent loop — nasıl oturacağına dair alınan kararları ve bunların getirdiği ek/değişen görevleri toplar. Yukarıdaki Faz 1-4'ü **iptal etmez**; kısıtlar ve netleştirir.

İlgili referanslar: `DOC-02` (RAG indeksleme), `DOC-03` (agent loop), `DOC-04` (citation doğrulama), `query_analyzer.py` (#179/#187), `citation_verifier.py`, `query_context.py`, `chunk_store.py`, `vector_index.py`.

### Karar 1 — Tree granülerliği: PER-DOCUMENT (kesin)

RAPTOR tree'leri **doküman başına** kurulur. Corpus-global, erişim-bandı-bazlı (corpus-profil) ve kullanıcı-başına tree seçenekleri **reddedildi**.

Gerekçe:
- **ACL (belirleyici):** `access_level` doküman düzeyinde uniform — `vector_index.py:126` "All chunks must share the same `doc_id` and `access_level`"; kaynağı `config.py:59` WATCHED_DIRS `path:access_level` (subpath'ten türetme yasak, açık olmalı). Per-doc özet, dokümanın seviyesini otomatik miras alır. Dokümanları karıştıran herhangi bir tree (global/bant) için özete atanacak **tek bir doğru `access_level` yoktur** → ya `confidential` sızıntısı ya recall kaybı. `models/access.py` "filtreyi unutmak mimari olarak imkânsız" garantisinde bu yapısal bir delik olur.
- **Staleness sınırlı:** doküman değişince yalnız o ağaç yeniden kurulur; per-doc content-hash reindex'e (`DOC-02 §2.7`) temiz bağlanır.
- **Citation provenance temiz:** özetin kaynakları tek `doc_id` → verbatim layer-0'a iniş tek doküman içinde.
- **Maliyet incremental + paralel:** rebuild **günlük churn** ile ölçeklenir (corpus boyutuyla değil); dokümanlar bağımsız olduğu için paralelleşir/sharding'e uygun.
- **Referans:** RAGFlow RAPTOR zaten per-doc (doküman parse task'ı içinde) çalışıyor.

Per-user reddi: ACL **banda dayalı, kullanıcıya değil**; kullanıcılar bandları paylaşır → per-user tree, "per-band × kullanıcı sayısı" kadar redundant (aynı içerik N kez). Doğru granülerlik bant sayısı (3), kullanıcı sayısı (N) değil — ve per-doc bunu da kapsar.

- [ ] Issue scope'unu "**per-document** RAPTOR" olarak netleştir.
- [ ] **Açık yasak:** corpus-global veya erişim-bandı-bazlı özet katmanı kurulmaz (ACL gerekçesi). İleride cross-doc precompute istenirse tek makul şekil "per-doc orman + bant başına ince **kanopi** (yalnız doküman-kök özetleri üzerinde, aynı `access_level` bandında)" — bu **v1 scope DIŞI**, gerekirse ayrı issue.

### Karar 2 — Motivasyon scope daraltma: per-doc, cross-doc DEĞİL

RAPTOR'un kazandırdığı yer: "bu dokümanın genel konusu ne?" ve **aynı uzun/yapısız dokümanın** farklı bölgelerini birleştiren tematik sorular. **Farklı dokümanlar arası** "X vs Y" hâlâ `gather_candidates → synthesize_structured` (COMPARISON/MULTI_DOC) yolundadır — RAPTOR bunu çözmez. Motivasyondaki "X ve Y konularını karşılaştır" örneği bu çerçeveyle daraltılmalı (aynı doküman içi farklı bölgeler).

- [ ] RAPTOR yalnızca **eşik üstü** dokümanda çalışsın: kısa dokümanlar `load_whole` + upload-time `description` ile zaten kapsanıyor; per-doc kök katmanı kısa dokümanda `description`'ı tekrarlar = boşa LLM çağrısı. Eşik için env: `RAG_RAPTOR_MIN_CHUNKS` (öneri ~20-30 chunk). Asıl değer paragraf-chunk ile whole-doc `description` **arasındaki orta katmanlarda**.

### Karar 3 — Citation: özetler routing-only, ASLA citation endpoint değil (KRİTİK)

Mevcut `citation_verifier.py` bir `raptor_summary` chunk'ına yapılan citation'ı **sessizce GEÇİRİR**: Step 1 retrieval_set ✓, Step 2 kendi byte_range'i ✓, Step 3 quote kendi sentetik metninde bulunur ✓. Yani naïf entegrasyon, makine-üretimi metne **audit-grade görünen** citation üretir — citation katmanının var olma sebebine (post-rationalized citation) doğrudan aykırı. Bunu engelleyen açık guard şart; "kendiliğinden" olmaz.

Kural: **cite edilen birim daima verbatim layer-0 chunk'tır.** Özet, "cevap nerede"yi bulan yönlendirme işaretçisidir; agent oradan layer-0'a inip alıntıyı oradan üretir.

- [ ] Özet chunk payload'ına `source_chunk_hashes` ekle — türediği layer-0 chunk'lara aşağı-işaretçi. `node_id`/`page` **değil** `chunk_hash` kullan (company_docs chunk'larında node_id/page null; chunk_hash baştan sona stabil anahtar).
- [ ] `ChunkMetadata`'ya (`chunk_store.py`) `raptor_layer` (veya `chunk_type`) alanı ekle; `QueryContext.add_retrieved_chunks` (`query_context.py:96`) doldursun. Verifier provenance görmeden guard yapamaz. Özet chunk'lar verbatim cite edilmeyeceği için byte_range null/sentinel olabilir (guard Step 2'den önce yakalar).
- [ ] `citation_verifier.py`: yeni `FailureType.synthetic_source`. Step 1'den sonra `chunk_meta.raptor_layer >= 1` ise verbatim segment için Step 2/3'e girmeden reddet; retry feedback: "chunk X üretilmiş bir özet; onun yerine kaynak chunk'larını cite et: [source_chunk_hashes]".
- [ ] İki citation rejimi netleşsin: **(a) verbatim** (default, 3-step) atomik fakt için — tek byte_range+quote hakkı olan tek rejim; **(b) per-source attribution** (`synthesize_structured`'ın `per_source_attributions`'ı) SUMMARY/overview segment için — özet metninde quote-match aranmaz, kaynak küme retrieval_set'te + her kaynak layer-0 doğrulanır.
- [ ] `synthesize_structured` da özet chunk'a değil **layer-0 kaynaklarına** attribute etsin (aynı provenance-expansion tool içinde).
- [ ] Test: özete yapılan verbatim citation `synthetic_source` ile reddediliyor; content-unavailable yolundan (`citation_verifier.py:191-198`) da sızamıyor.

### Karar 4 — Güncelleme/staleness: content-hash'e bağla (Faz 2'yi netleştirir)

- [ ] RAPTOR (re)build tetikleyicisi = mevcut content-hash change detection (`DOC-02 §2.7`). `status="unchanged"` → RAPTOR'u **atla** (tekrar LLM ödeme yok); `status="reindexed"` → eski raptor chunk'larını sil (`delete_points_by_doc_id` zaten kapsıyor) + yeniden kur.
- [ ] Özet chunk payload'ına `raptor_built_from_hash` (kaynak layer-0 içerik hash'i) yaz; dokümanın güncel hash'i ile uyuşmazsa özet **bayat** → retrieval'da gizle / rebuild işaretle. (Sessizce bayat özet servis etmeyi önler.)
- [ ] Test (atomiklik): değişmiş dokümanı reindex et → eski `raptor_summary` point'leri gitti, yenileri var, `count > 0`.

### Karar 5 — Maliyet & zamanlama: upload kritik yolundan ayır, gece batch

RAPTOR LLM çağrıları lokal GPU'da koşar → **per-token dolar maliyeti yok**; maliyet = GPU-zamanı + index gecikmesi. Asıl maliyet sürücüsü tek-seferlik build değil, **rebuild sıklığı × tree kapsamı** — per-doc'ta bu günlük churn kadar (sınırlı). Gece penceresi sonlu olduğundan incremental+paralel per-doc bu modele oturur.

- [ ] Upload/index gündüz hızlı kalsın: yalnız chunk + embedding (LLM özetleme yok). `RAG_RAPTOR_ENABLED=false` upload yolunda default kalır.
- [ ] RAPTOR'u **scheduled gece batch** olarak çalıştır (boş GPU): o gün hash'i değişen dokümanların ağaçlarını (yeniden) kur. Dokümanlar bağımsız → paralel/sharded.
- [ ] Tek seferlik ilk kurulum = **hafta sonu backfill** olarak ele al (corpus boyutuyla orantılı, bir kez ödenir).

### Karar 6 — Tool yüzeyi: RAPTOR tool-görünmez kalır

RAPTOR, agent'ın 7-tool yüzeyine **hiçbir şey eklemez** (index-time artifact; özet chunk'lar `semantic_search` sonuçlarında şeffaf gelir). "Az tool / LLM için basit karar" prensibine dokunmaz — bu eksen RAPTOR'la ilgisiz.

- [ ] `raptor_search` / `summarize_document` gibi **yeni tool açılmaz**. Layer kontrolü gerekiyorsa retriever'da opsiyonel `raptor_layer` filtresi + query-class routing (`query_analyzer.py`, #179/#187) ile yapılır — yeni tool ile değil.
- [ ] Faz 3 routing doğru yönde (bu kararla teyitli): SUMMARY → `raptor_layer >= 1` tercih; LOOKUP/FACTUAL/KEYWORD → `raptor_layer == 0`.

### Karar 7 — Ship eval-gated kalır

- [ ] Default off; ship kararı **ablation'a bağlı**: RAPTOR-on, mevcut SUMMARY yolunu (`get_outline + chunk_read + synthesize_structured` + doküman `description`) doküman başına N LLM çağrısı + index şişmesi + routing karmaşasını haklı çıkaracak kadar geçiyor mu? Eval'i **yapısız/uzun dokümanlarda** kur (RAPTOR'un getirisi orada; iyi-TOC'lu dokümanda marjinal). Mecra: #223 (parametre kalibrasyonu) + eval set.
